dek-registry/src/main/java/io/confluent/dekregistry/storage/DekRegistry.java

```diff
@@ -759,31 +792,325 @@ private void deleteDek(String name, String subject, DekFormat algorithm, boolean
-    DataEncryptionKey key = (DataEncryptionKey) keys.get(keyId);
+    DataEncryptionKeyId id = new DataEncryptionKeyId(tenant, name, subject, version, algorithm);
+    DataEncryptionKey key = (DataEncryptionKey) keys.get(id);
     if (key == null) {
       return;
     }
     if (permanentDelete) {
       if (!key.isDeleted()) {
         throw new KeyNotSoftDeletedException(subject);
       }
-      keys.remove(keyId);
+      keys.remove(id);
     } else {
       if (!key.isDeleted()) {
         DataEncryptionKey newKey = new DataEncryptionKey(name, key.getSubject(),
             key.getVersion(), key.getAlgorithm(), key.getEncryptedKeyMaterial(), true);
-        keys.put(keyId, newKey);
+        keys.put(id, newKey);
       }
     }
   }
 
+  public void undeleteKekOrForward(String name, Map<String, String> headerProperties)
+      throws SchemaRegistryException {
+    String tenant = schemaRegistry.tenant();
+    lock(tenant, headerProperties);
+    try {
+      if (isLeader(headerProperties)) {
+        undeleteKek(name);
+      } else {
+        // forward registering request to the leader
+        if (schemaRegistry.leaderIdentity() != null) {
+          forwardUndeleteKekRequestToLeader(name, headerProperties);
+        } else {
+          throw new UnknownLeaderException("Request failed since leader is unknown");
+        }
+      }
+    } finally {
+      unlock(tenant);
+    }
+  }
+
+  private void forwardUndeleteKekRequestToLeader(String name, Map<String, String> headerProperties)
+      throws SchemaRegistryRequestForwardingException {
+    RestService leaderRestService = schemaRegistry.leaderRestService();
+    final UrlList baseUrl = leaderRestService.getBaseUrls();
+
+    UriBuilder builder = UriBuilder.fromPath("/dek-registry/v1/keks/{name}/undelete");
+    String path = builder.build(name).toString();
+
+    log.debug(String.format("Forwarding undelete key request to %s", baseUrl));
+    try {
+      leaderRestService.httpRequest(
+          path, "POST", null, headerProperties, VOID_TYPE);
+    } catch (IOException e) {
+      throw new SchemaRegistryRequestForwardingException(
+          String.format("Unexpected error while forwarding the undelete key request to %s",
+              baseUrl),
+          e);
+    } catch (RestClientException e) {
+      throw new RestException(e.getMessage(), e.getStatus(), e.getErrorCode(), e);
+    }
+  }
+
+  private void undeleteKek(String name) throws SchemaRegistryException {
+    // Ensure cache is up-to-date
+    keys.sync();
+
+    String tenant = schemaRegistry.tenant();
+    EncryptionKeyId keyId = new KeyEncryptionKeyId(tenant, name);
+    KeyEncryptionKey key = (KeyEncryptionKey) keys.get(keyId);
+    if (key == null) {
+      return;
+    }
+    if (key.isDeleted()) {
+      KeyEncryptionKey newKey = new KeyEncryptionKey(name, key.getKmsType(),
+          key.getKmsKeyId(), key.getKmsProps(), key.getDoc(), key.isShared(), false);
+      keys.put(keyId, newKey);
+    }
```

Comment:
What if key being undeleted is not soft deleted yet? Shouldnt we return an error?

Reply:
Good question.  The undelete is used to recover some soft-deleted items.  Imagine if they have a mix of DEKs, some soft-deleted and some not.  If they are want to recover all DEKs for a subject, it would be inconvenient to make them soft-delete the ones they want to recover before they can recover all of them.  So we allow recovery to be more lenient, it doesn't complain if objects are not soft-deleted.